Test, CMake, CI

kgerheiser opened the pull request.

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+  string(TOUPPER ${kind} kind_definition)
+
+  set(BUILD_FLAGS "${fortran_${kind}_flags}")
+  if(OpenMP_Fortran_FOUND)
+    set(BUILD_FLAGS "${BUILD_FLAGS} ${OpenMP_Fortran_FLAGS}")
+  endif()
```
I think we should do this instead.
remove lines 25-27 and replace with:
```
if(OpenMP_Fortran_FOUND)
  target_link_libraries(${lib_name} OpenMP::OpenMP_Fortran)
endif()
```
before the `target_include_directories` line below.

kgerheiser commented:
Yes.

I create `main.yml` which builds and runs `make test` on every PR/commit. You can see the results of that here (it built and ran after I pushed to this branch) https://github.com/kgerheiser/NCEPLIBS-ip/runs/759818696?check_suite_focus=true


I move the tests into the `tests/` directory and converted the tests to use CMake/pFUnit


Some CMake things were moved around, or changed a little. I took the version number out of the library name, and added a namespace to the imported target.

aerorahul commented on cmake/PackageConfig.cmake.in:
```diff
@@ -0,0 +1,12 @@
+
+# Include targets file.  This will create IMPORTED target @PROJECT_NAME@
+include("${CMAKE_CURRENT_LIST_DIR}/@PROJECT_NAME@-targets.cmake")
+
```
we are adding a dependency on OpenMP, so that should be found if this library was compiled with that option.
Something like this:
```
# ON/OFF implies ip was compiled with/without OpenMP
if(@OpenMP_Fortran_FOUND@)
  find_dependency(OpenMP)
endif()
```
The documentation is [here](https://cmake.org/cmake/help/latest/module/CMakeFindDependencyMacro.html).
It says there is no need to pass arguments such as `REQUIRED` or `COMPONENTS Fortran` as they are passed from the original call to `find_package`, but we should check that is correctly done.

GeorgeGayno-NOAA commented:
Can you compile/run the tests locally?

aerorahul commented:
> Can you compile/run the tests locally?
@GeorgeGayno-NOAA 
yes, if one builds with `ENABLE_TEST` and have those dependencies `pFunit`.
```
$> ctest
```

GeorgeGayno-NOAA commented:
> Can you compile/run the tests locally?
> @GeorgeGayno-NOAA
> yes, if one builds with ENABLE_TEST and have those dependencies pFunit.
> 
> $> ctest

The README file in the ./tests directory has a section on how to compile and run the tests.  That section should be updated to reflect this new method.

aerorahul commented on README.md:
````diff
@@ -0,0 +1,29 @@
+make -j2
+make install
+```
+
````
```
# Testing

Testing requires [pFUnit](https://github.com/Goddard-Fortran-Ecosystem/pFUnit).
```

```
cmake .. -DENABLE_TESTS=ON -DCMAKE_PREFIX_PATH="/path/to/pfunit /path/to/NCEPLIBS"
make -j2
ctest
```

GeorgeGayno-NOAA replied:
Ok.  That is good.  

There is another readme in the ./tests subdirectory.  You should remove the section on how to run the test (now obsolete) or refer the user to the README.md.

GeorgeGayno-NOAA replied:
Having some trouble building it.  

Another question, does test_mod.pf replace the tests in ./unit_test and ./unit_test_for_c?

kgerheiser replied:
Yes, `test_mod.pf` (should probably be given a better name) contains the code for each unit test. I re-factored the code from those files and converted it to pfunit. I did the scalar part but not the vector. After I finish the other libraries I'll come back and add the vector tests.

aerorahul commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+  string(TOUPPER ${kind} kind_definition)
+
+  set(BUILD_FLAGS "${fortran_${kind}_flags}")
+  if(OpenMP_Fortran_FOUND)
+    target_link_libraries(${lib_name} OpenMP::OpenMP_Fortran)
+  endif()
```
This section needs to move after the `add_library()` line

aerorahul commented:
@kgerheiser One final find_dependency(OpenMP) in the PackageConfig.cmake.in is needed before this can be merged.  See [this](https://github.com/NOAA-EMC/NCEPLIBS-ip/pull/10#discussion_r438792539) comment.

aerorahul commented on cmake/PackageConfig.cmake.in:
```diff
@@ -2,10 +2,15 @@
 # Include targets file.  This will create IMPORTED target @PROJECT_NAME@
 include("${CMAKE_CURRENT_LIST_DIR}/@PROJECT_NAME@-targets.cmake")
 
+if(@OpenMP_Fortran_FOUND@)
```
Can you retain the comment line above this if-block.  Otherwise, when this is resolved it will look like this:
```
if(ON)
  find_dependency(OpenMP)
endif()
```

aerorahul replied:
replacing  `if(@OpenMP_Fortran_FOUND@)` with `if(@OPENMP@)` works.

aerorahul commented:
@kgerheiser There is a problem when not building with OpenMP in the package Config

kgerheiser commented:
So, put `if(@OPENMP@)` in the package config `if`?

aerorahul commented:
> So, put `if(@OPENMP@)` in the package config `if`?

yes. I pushed a bugfix to the commit
